hpx/parallel/spmd_block.hpp

```diff
@@ -0,0 +1,156 @@
+    /// to each image. The constraint for the lambda given to the
+    /// define_spmd_block function is to accept a spmd_block as first parameter.
+    struct spmd_block
+    {
+        spmd_block(){}
+
+        spmd_block(
+            std::string name, std::size_t num_images, std::size_t image_id)
+        : name_(name), num_images_(num_images), image_id_(image_id)
+        {}
+
+        std::size_t get_num_images() const
+        {
+            return num_images_;
+        }
+
+        std::size_t this_image() const
+        {
+            return image_id_;
+        }
+
+    // FIXME : If 2 images are scheduled in the same thread (not OS-thread)
+    // -> deadlock
+        void sync_all() const
+        {
+           if (!barrier_)
+           {
+               barrier_ = std::make_shared<hpx::lcos::barrier>(
+                 name_ + "_barrier"
+               , num_images_
+               );
+           }
+
+           barrier_->wait();
+        }
+
+        hpx::future<void> sync_all(hpx::launch::async_policy const &) const
+        {
+           if (!barrier_)
+           {
+               barrier_ = std::make_shared<hpx::lcos::barrier>(
+                 name_+ "_barrier"
+               , num_images_
+               );
+           }
+
+           return barrier_->wait(hpx::launch::async);
+        }
+
+    private:
+        std::string name_;
+        std::size_t num_images_;
+        std::size_t image_id_;
+        mutable std::shared_ptr<hpx::lcos::barrier> barrier_;
+
+/*    private:
+        friend class hpx::serialization::access;
+
+        template <typename Archive>
+        void serialize(Archive& ar, unsigned int const)
+        {
+            ar & name_ & num_images_ && image_id_;
+        }*/
+    };
+
+    template <typename F, typename ... Args>
+    void define_spmd_block(
+        std::string && name,
+        std::size_t num_images,
+        F && f, Args && ... args)
+    {
+        using hpx::parallel::execution::par;
+        using ftype = typename std::remove_reference<F>::type;
+
+        using first_type
+            = typename
+                hpx::parallel::v2::detail::extract_first_parameter<
+                    decltype(&ftype::operator())>::type;
+
+        static_assert( std::is_same<spmd_block,first_type>::value,
+            "define_spmd_block() needs a lambda that " \
+            "has at least a spmd_block as 1st argument");
+
+        auto range = boost::irange(0ul, num_images);
+        hpx::parallel::static_chunk_size cs(1);
+
+// FIXME : How to invoke images remotely?
+
+/*        auto a = hpx::actions::lambda_to_action(std::move(f));*/
+
+    // Note : par.with() ensures that each invoked image is running in a
+    // separate thread
+        hpx::parallel::for_each(
+            par.with(cs), range.begin(), range.end(),
+            [=,&f](std::size_t image_id)
+            {
+                spmd_block block(name,num_images,image_id);
+                f(block,args...);
+            });
```

Comment:
Similarly, you could enable to use execution policies (instead of the executor) which had the advantage to seamlessly enable an asynchronous `define_spmd_block` construct:

    future<void> f = define_spmd_block(par(task), ...);

The implementation would be almost identical. See the implementation of `task_block` on how this could be done (https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/parallel/task_block.hpp)